src/core/data/ndarray/ndarray_vector.h

```diff
@@ -5,12 +5,216 @@
+            : nCells_{ncells}
+            , data_(std::accumulate(ncells.begin(), ncells.end(), 1, std::multiplies<int>()))
+        {
+        }
+
+
+        NdArrayVector(NdArrayVector const& source) = default;
+        NdArrayVector(NdArrayVector&& source)      = default;
+
+        static constexpr bool is_contiguous = 1;
+
+        auto data() const { return data_.data(); }
+
+        auto size() const { return data_.size(); }
+
+        auto begin() const { return std::begin(data_); }
+        auto begin() { return std::begin(data_); }
+
+        auto end() const { return std::end(data_); }
+        auto end() { return std::end(data_); }
+
+        void zero() { data_ = std::vector<DataType>(data_.size(), {0}); }
+
+
+        NdArrayVector& operator=(NdArrayVector const& source)
+        {
+            if (nCells_[0] != source.nCells_[0])
+            {
+                throw std::runtime_error(
+                    "Error NdArrayVector cannot be assigned, incompatible sizes");
+            }
+
+            if constexpr (dim > 1)
+            {
+                if (nCells_[1] != source.nCells_[1])
+                {
+                    throw std::runtime_error(
+                        "Error NdArrayVector cannot be assigned, incompatible sizes");
+                }
+
+                if constexpr (dim > 2)
+                {
+                    if (nCells_[2] != source.nCells_[2])
+                    {
+                        throw std::runtime_error(
+                            "Error NdArrayVector cannot be assigned, incompatible sizes");
+                    }
+                }
+            }
+
+            this->data_ = source.data_;
+            return *this;
+        }
+
+        NdArrayVector& operator=(NdArrayVector&& source)
+        {
+            if (nCells_[0] != source.nCells_[0])
+            {
+                throw std::runtime_error(
+                    "Error NdArrayVector cannot be assigned, incompatible sizes");
+            }
+
+            if constexpr (dim > 1)
+            {
+                if (nCells_[1] != source.nCells_[1])
+                {
+                    throw std::runtime_error(
+                        "Error NdArrayVector cannot be assigned, incompatible sizes");
+                }
+
+                if constexpr (dim > 2)
+                {
+                    if (nCells_[2] != source.nCells_[2])
+                    {
+                        throw std::runtime_error(
+                            "Error NdArrayVector cannot be assigned, incompatible sizes");
+                    }
+                }
+            }
+
+            this->data_ = std::move(source.data_);
+            return *this;
+        }
+
+
+        template<typename... Indexes>
+        DataType& operator()(Indexes... indexes)
+        {
+            auto params = std::tuple<Indexes...>{indexes...};
+            static_assert(sizeof...(Indexes) == dim);
```

Comment:
Could replace the op per dim

```
static_assert((... && std::is_unsigned_v<decltype(indexes)>));
```

Reply:
I think having (Indexes... indexes) with the static_assert that checks we have dim arguments will break @Dekken PR #230

Reply:
I'm going to dry the Index/sub_array thing so it should be ok

Reply:
I'm not that sure to understand the suggestion ; and it appears to me less talkative. but if it's a better solution, you guys will have to explain it to me.

Reply:
hmmm, I will need your help on this one : these comments are not explicit to me.